### PR TITLE
[BEAM-5852] BeamSQL functions

### PR DESCRIPTION
**Pull request by amaliujia**
1. Add some small string functions.
2. Update the way to test builtin functions.
3. Add NULL check for builtin functions.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

File: sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/impl/udf/BeamSalUhfSpecialTypeAndValueTest.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.Row;
+import org.apache.commons.lang3.ArrayUtils;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Unit tests for UDFs. */
+@RunWith(JUnit4.class)
+public class BeamSalUhfSpecialTypeAndValueTest extends BeamSqlDslBase {
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/impl/udf/BeamSalUhfSpecialTypeAndValueTest.java]: Some special values (e.g. +/- INF) cannot be tested by `SELECT func(INF)`. So such values are tested by mocking table approach.

**Comment by amaliujia** [comment]
@apilloud @kennknowles @akedin

**Comment by akedin** [comment]
LGTM

**Comment by akedin** [comment]
run java precommit

**Comment by amaliujia** [comment]
Thanks!

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/udf/BuiltinStringFunctions.java [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    funcName = "STARTS_WITH",
+    parameterArray = {TypeName.STRING},
+    returnType = TypeName.STRING
+  )
+  public Boolean startsWith(String str1, String str2) {
+    return str1.startsWith(str2);
+  }
+
+  @UDF(
+    funcName = "LENGTH",
```
Comment (apilloud) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/udf/BuiltinStringFunctions.java]: This looks to be identical to CHARACTER_LENGTH in calcite?

Comment (amaliujia) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/udf/BuiltinStringFunctions.java]: As the reference, [functions-and-operators#length](https://cloud.google.com/bigquery/docs/reference/standard-sql/functions-and-operators#length), `length` handles `character array` and `string`. In calcite, CHARACTER_LENGTH only handle `string` (https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java#L235).


In BigQuery, `string` is defined as UTF8 encoded character array. Functions handle string and character array differently. See: https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#string-type

Comment (amaliujia) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/udf/BuiltinStringFunctions.java]: You can go ahead to use code generation. For any function that we might want them behave differently, we can implement those and let planner decides which set of functions to use.

**Pull request merged**